src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+            }
+
+            using var region = new PlaceholderRegion(this, placeholders);
+            placeholders.Free();
+
+            return base.VisitUsingStatement(node);
+        }
+
+        public override BoundNode? VisitUsingLocalDeclarations(BoundUsingLocalDeclarations node)
+        {
+            var placeholders = ArrayBuilder<(BoundValuePlaceholderBase, uint)>.GetInstance();
+            if (node.AwaitOpt is { } awaitableInfo)
+            {
+                GetAwaitableInstancePlaceholders(placeholders, awaitableInfo, _localScopeDepth);
+            }
+
+            using var region = new PlaceholderRegion(this, placeholders);
+            placeholders.Free();
+
+            return base.VisitUsingLocalDeclarations(node);
+        }
+
+        public override BoundNode? VisitFixedStatement(BoundFixedStatement node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitFixedStatement(node);
+        }
+
+        public override BoundNode? VisitDoStatement(BoundDoStatement node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitDoStatement(node);
+        }
+
+        public override BoundNode? VisitWhileStatement(BoundWhileStatement node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitWhileStatement(node);
+        }
+
+        public override BoundNode? VisitSwitchStatement(BoundSwitchStatement node)
+        {
+            using var _ = new LocalScope(this, node.InnerLocals);
+            using var patternInput = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            base.VisitSwitchStatement(node);
+            return null;
+        }
+
+        public override BoundNode? VisitConvertedSwitchExpression(BoundConvertedSwitchExpression node)
+        {
+            using var patternInput = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            base.VisitConvertedSwitchExpression(node);
+            return null;
+        }
+
+        public override BoundNode? VisitSwitchSection(BoundSwitchSection node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchSection(node);
+        }
+
+        public override BoundNode? VisitSwitchExpressionArm(BoundSwitchExpressionArm node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchExpressionArm(node);
+        }
+
+        public override BoundNode? VisitCatchBlock(BoundCatchBlock node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitCatchBlock(node);
+        }
+
+        public override BoundNode? VisitLocal(BoundLocal node)
+        {
+            Debug.Assert(_localEscapeScopes?.ContainsKey(node.LocalSymbol) == true ||
```

Comment:
`_localEscapeScopes` is lazily created when `AddLocalScopes()` is called, so in unexpected cases (when this assert might fail for instance), the value might be `null`.

Reply:
I think we should have called `AddLocalScopes()` in all scenarios before calling `VisitLocal`. So I would not do the `null` check here. The exception seems desirable.

Reply:
`_localEscapeScopes` is only used in the assert though. It looks like the only difference is how the assert fails: because the condition is `false` or because of a `NullReferenceException`.

Reply:
Actually, `_localEscapeScopes` may be `null` for a local in a top-level statement.